### Import Package

In [249]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

### Import Data 'data_auto.csv'

In [250]:
df = pd.read_csv("data_auto.csv", index_col='Unnamed: 0')
df.head()

cylinders  displacement horsepower  weight  acceleration  model_year  \
0          8         307.0      130.0  3504.0          12.0          70   
1          8         350.0      165.0  3693.0          11.5          70   
2          8         318.0      150.0  3436.0          11.0          70   
3          8         304.0      150.0  3433.0          12.0          70   
4          8         302.0      140.0  3449.0          10.5          70   

   origin                   car_name   mpg  
0       1  chevrolet chevelle malibu  18.0  
1       1          buick skylark 320  15.0  
2       1         plymouth satellite  18.0  
3       1              amc rebel sst  16.0  
4       1                ford torino  17.0

mengambil index kolom 'Unnamed: 0' sebagai index pada dataset

### Melihat jumlah baris dan kolom

In [251]:
df.shape

(398, 9)

terdapat 398 baris dan 9 kolom<br>
terdapat 8 kolom fitur dan 1 kolom target 

### Melihat data yang kosong

In [252]:
df.replace("?", np.nan, inplace = True)
df.head()

cylinders  displacement horsepower  weight  acceleration  model_year  \
0          8         307.0      130.0  3504.0          12.0          70   
1          8         350.0      165.0  3693.0          11.5          70   
2          8         318.0      150.0  3436.0          11.0          70   
3          8         304.0      150.0  3433.0          12.0          70   
4          8         302.0      140.0  3449.0          10.5          70   

   origin                   car_name   mpg  
0       1  chevrolet chevelle malibu  18.0  
1       1          buick skylark 320  15.0  
2       1         plymouth satellite  18.0  
3       1              amc rebel sst  16.0  
4       1                ford torino  17.0

In [253]:
df.isnull().any()

cylinders       False
displacement    False
horsepower       True
weight          False
acceleration    False
model_year      False
origin          False
car_name        False
mpg             False
dtype: bool

In [254]:
df.isnull().sum()

cylinders       0
displacement    0
horsepower      6
weight          0
acceleration    0
model_year      0
origin          0
car_name        0
mpg             0
dtype: int64

terdapat 6 data yang kosong pada kolom horsepower

### Melihat tipe data

In [255]:
df.dtypes

cylinders         int64
displacement    float64
horsepower       object
weight          float64
acceleration    float64
model_year        int64
origin            int64
car_name         object
mpg             float64
dtype: object

### Merubah tipe data sesuai dengan kolom dan deskripsi

In [256]:
df['horsepower'] = df['horsepower'].astype(float)
df.dtypes

cylinders         int64
displacement    float64
horsepower      float64
weight          float64
acceleration    float64
model_year        int64
origin            int64
car_name         object
mpg             float64
dtype: object

tipe data pada setiap kolom sudah sesuai dengan deskripsi

### Melihat deskripsi setiap kolom

In [257]:
df.describe()

cylinders  displacement  horsepower       weight  acceleration  \
count  398.000000    398.000000  392.000000   398.000000    398.000000   
mean     5.454774    193.425879  104.469388  2970.424623     15.568090   
std      1.701004    104.269838   38.491160   846.841774      2.757689   
min      3.000000     68.000000   46.000000  1613.000000      8.000000   
25%      4.000000    104.250000   75.000000  2223.750000     13.825000   
50%      4.000000    148.500000   93.500000  2803.500000     15.500000   
75%      8.000000    262.000000  126.000000  3608.000000     17.175000   
max      8.000000    455.000000  230.000000  5140.000000     24.800000   

       model_year      origin         mpg  
count  398.000000  398.000000  398.000000  
mean    76.010050    1.572864   23.514573  
std      3.697627    0.802055    7.815984  
min     70.000000    1.000000    9.000000  
25%     73.000000    1.000000   17.500000  
50%     76.000000    1.000000   23.000000  
75%     79.000000    2.000000   29.000000  
max     82.000000    3.000000   46.600000

### Mengisi data yang kosong pada kolom horsepower dengan mengambil nilai rata-rata

In [258]:
df.fillna(df['horsepower'].mean(), inplace = True)
df.head()

cylinders  displacement  horsepower  weight  acceleration  model_year  \
0          8         307.0       130.0  3504.0          12.0          70   
1          8         350.0       165.0  3693.0          11.5          70   
2          8         318.0       150.0  3436.0          11.0          70   
3          8         304.0       150.0  3433.0          12.0          70   
4          8         302.0       140.0  3449.0          10.5          70   

   origin                   car_name   mpg  
0       1  chevrolet chevelle malibu  18.0  
1       1          buick skylark 320  15.0  
2       1         plymouth satellite  18.0  
3       1              amc rebel sst  16.0  
4       1                ford torino  17.0

In [259]:
df.isnull().any()

cylinders       False
displacement    False
horsepower      False
weight          False
acceleration    False
model_year      False
origin          False
car_name        False
mpg             False
dtype: bool

sudah tidak terdapat data yang kosong pada masing-masing kolom

### Melakukan Label Encoding pada kolom kategorikal

In [260]:
df['cylinders'].value_counts()

4    204
8    103
6     84
3      4
5      3
Name: cylinders, dtype: int64

cylinders merupakan kategorikal

In [261]:
df['model_year'].value_counts()

73    40
78    36
76    34
82    31
75    30
70    29
79    29
80    29
81    29
71    28
72    28
77    28
74    27
Name: model_year, dtype: int64

model_year merupakan kategorikal

In [262]:
df['origin'].value_counts()

1    249
3     79
2     70
Name: origin, dtype: int64

origin merupakan kategorikal

In [263]:
df['car_name'].value_counts()

ford pinto             6
toyota corolla         5
amc matador            5
ford maverick          5
chevrolet chevette     4
                      ..
chevrolet monza 2+2    1
ford mustang ii        1
pontiac astro          1
amc pacer              1
chevy s-10             1
Name: car_name, Length: 305, dtype: int64

car_name merupakan kategorikal

In [272]:
from sklearn.preprocessing import LabelEncoder

encoding = LabelEncoder()
df['cylinders'] = encoding.fit_transform(df['cylinders'])
df['model_year'] = encoding.fit_transform(df['model_year'])
df['origin'] = encoding.fit_transform(df['origin'])
df['car_name'] = encoding.fit_transform(df['car_name'])
df.head()

cylinders  displacement  horsepower    weight  acceleration  model_year  \
0          4      0.617571    0.456522  0.536150      0.238095           0   
1          4      0.728682    0.646739  0.589736      0.208333           0   
2          4      0.645995    0.565217  0.516870      0.178571           0   
3          4      0.609819    0.565217  0.516019      0.238095           0   
4          4      0.604651    0.510870  0.520556      0.148810           0   

   origin  car_name   mpg  
0       0        49  18.0  
1       0        36  15.0  
2       0       231  18.0  
3       0        14  16.0  
4       0       161  17.0

melakukan proses encoding untuk data fitur berupa teks ataupun kategorikal<br>
agar dapat diproses oleh mesin serta memudahkan dalam melakukan labeling pada data kategorikal

### Merubah tipe data kolom car_name menyesuaikannya dengan isi datanya yaitu integer

In [265]:
df.dtypes

cylinders         int64
displacement    float64
horsepower      float64
weight          float64
acceleration    float64
model_year        int64
origin            int64
car_name          int32
mpg             float64
dtype: object

In [266]:
df['car_name'] = df['car_name'].astype(int)
df.dtypes

cylinders         int64
displacement    float64
horsepower      float64
weight          float64
acceleration    float64
model_year        int64
origin            int64
car_name          int32
mpg             float64
dtype: object

tipe data sudah sesuai untuk semua kolom

### Melakukan normalisasi menggunakan MinMax Scaler untuk kolom numerik

In [267]:
df.head()

cylinders  displacement  horsepower  weight  acceleration  model_year  \
0          4         307.0       130.0  3504.0          12.0           0   
1          4         350.0       165.0  3693.0          11.5           0   
2          4         318.0       150.0  3436.0          11.0           0   
3          4         304.0       150.0  3433.0          12.0           0   
4          4         302.0       140.0  3449.0          10.5           0   

   origin  car_name   mpg  
0       0        49  18.0  
1       0        36  15.0  
2       0       231  18.0  
3       0        14  16.0  
4       0       161  17.0

In [268]:
from sklearn.preprocessing import MinMaxScaler
numeric = ["displacement", "horsepower", "weight", "acceleration"]
scaler = MinMaxScaler(feature_range=(0, 1))

scaler.fit(df.loc[:, df[numeric].columns])
df.loc[:, df[numeric].columns] = scaler.transform(df.loc[:, df[numeric].columns])
df.head()

cylinders  displacement  horsepower    weight  acceleration  model_year  \
0          4      0.617571    0.456522  0.536150      0.238095           0   
1          4      0.728682    0.646739  0.589736      0.208333           0   
2          4      0.645995    0.565217  0.516870      0.178571           0   
3          4      0.609819    0.565217  0.516019      0.238095           0   
4          4      0.604651    0.510870  0.520556      0.148810           0   

   origin  car_name   mpg  
0       0        49  18.0  
1       0        36  15.0  
2       0       231  18.0  
3       0        14  16.0  
4       0       161  17.0

melakukan normalisasi minmax untuk semua kolom fitur numerikal 

### Dataset Splitting

In [269]:
X = df.drop(columns=["mpg"])
y = df["mpg"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

tidak memakai stratify karena bukan merupakan klasifikasi dan kolom target berupa numerikal<br>
data training 80%, data testing 20%

### Fitting Dataset menggunakan Linear Regression

In [270]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_train, y_train)

print('Training Score\t= ', model.score(X_train, y_train), '\nTesting Score\t= ', model.score(X_test, y_test))

Training Score	=  0.8260787885067236 
Testing Score	=  0.8192819177155614


### Melihat nilai Evaluasi data

In [271]:
hasil = model.predict(X_test)
from sklearn.metrics import mean_squared_error, r2_score

print('MSE\t= ', mean_squared_error(y_test, hasil), '\nR2\t= ', r2_score(y_test, hasil))

MSE	=  11.467895116804481 
R2	=  0.8192819177155614
